### PR TITLE
Keycloak user management

### PR DESCRIPTION
**Pull request by MancunianSam**
- Set up keycloak users
- Add github actions

### CODE REVIEW

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    .serverUrl(auth.url)
+    .realm("tdr")
+    .clientId(auth.client)
+    .clientSecret(auth.secret)
+    .grantType(OAuth2Constants.CLIENT_CREDENTIALS)
+    .build()
+
+  def deleteUser(auth: Auth, userId: String): String = {
+    val client = keyCloakAdminClient(auth)
+    val realm = client.realm("tdr")
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala:27]: Is it worth passing in the realm as an environment variable. Unlikely as it is, we may one need to do a bulk upload of users to the `master` realm, or another realm? Think this should a least be set as a `val` as it is used multiple times

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/LambdaRunner.scala [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import scala.jdk.CollectionConverters._
+
+object LambdaRunner extends App {
+
+  sendCsvLambdaRequest()
+  sendApiLambdaRequest()
+
+  def sendCsvLambdaRequest() = {
+    // You will need to run this with credentials which have access to this bucket in the sandbox account
+    val s3Entity = new S3Entity(null, new S3BucketEntity("tdr-create-bulk-keycloak-users-intg", null, null), new S3ObjectEntity("sample_csv_file.csv", null, null, null, null), null)
```
Comment (TomJKing) [inline src/main/scala/uk/gov/nationalarchives/keycloak/users/LambdaRunner.scala:18]: Should the bucket entity name be hardcoded? The bucket looks like its using the one in integration? Should this actually be the sandbox one, as the comment suggests?

File: src/test/resources/application.conf [hunk 3]
```diff
@@ -0,0 +1,17 @@
+auth {
+  secret = "YQo="
+  url = "aHR0cDovL2xvY2FsaG9zdDo5MDAyL2F1dGg=" # http://localhost:9002/auth
```
Comment (TomJKing) [inline src/test/resources/application.conf:3]: Maybe update the comment to read: `Base 64 decodes to this value to http://localhost:9002/auth`

**Review by TomJKing — COMMENTED** [review]
Looks good. Just some minor questions and suggestions.

The `README` should be updated

Had a look at the repo set up:

- `Automatically delete head branches` option needs to be set

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  push:
+    branches:
+      - main
+permissions:
+  id-token: write
+  contents: write
+jobs:
+  pre-deploy:
+    runs-on: ubuntu-latest
+    environment: intg
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply .github/workflows/build.yml:12]: Yeah, this is building the lambda jar and uploading it to the backend code s3 bucket. It's equivalent to this bit https://github.com/nationalarchives/tdr-jenkinslib/blob/master/vars/lambdaTests.groovy#L50 in the Jenkinsfile and that has intg hardcoded.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          message: '${{ steps.bump-semver.outputs.new_version }}'
+      - name: Configure AWS credentials from management account
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          role-to-assume: arn:aws:iam::${{ secrets.MANAGEMENT_ACCOUNT }}:role/TDRGithubActionsRoleMgmt
+          aws-region: eu-west-2
+          role-session-name: APIDownloadDependencies
+      - name: Build jar file
+        run: sbt assembly
+      - name: Configure AWS credentials from integration account
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply .github/workflows/build.yml]: This shouldn't be here at all, we can do the upload to S3 using only the management account credentials.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          role-to-assume: arn:aws:iam::${{ secrets.ACCOUNT_NUMBER }}:role/TDRGithubActionsDeployLambda${{ secrets.TITLE_STAGE }}
+          aws-region: eu-west-2
+          role-session-name: APIDownloadDependencies
+      - name: Build new image version
+        env:
+          GITHUB_TOKEN: ${{ secrets.WORKFLOW_PAT }}
+        run: |
+          aws s3 cp target/scala-2.13/keycloak-user-management.jar s3://tdr-backend-checks-intg/${{ steps.bump-semver.outputs.new_version }}/keycloak-user-management.jar
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply .github/workflows/build.yml]: This is wrong as well. It should be going into the bucket in the management account to match the existing process. I've changed it.

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/CSVLambda.scala [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+  override def handleRequest(input: S3EventNotification, context: Context): String = {
+    for {
+      auth <- authFromConfig()
+      records <- IO(input.getRecords.asScala)
+      recordUsers <- IO(parseUsers(records))
+    } yield createUsers(auth, recordUsers)
+  }.unsafeRunSync()
+
+  private def parseUsers(records: mutable.Buffer[S3EventNotification.S3EventNotificationRecord]): List[UserCredentials] = {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/keycloak/users/CSVLambda.scala:31]: Makes sense, I've changed it.

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  def createUsers(auth: Auth, userCredentials: List[UserCredentials]): String = {
+    val client = keyCloakAdminClient(auth)
+    def createUser(userCredentials: UserCredentials) = {
+      val realm = client.realm("tdr")
+      val user = realm.users()
+
+
+      val userRepresentation: UserRepresentation = new UserRepresentation
+
+      val creds = userCredentials.password.map(password => {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala:43]: I've set it up so you can directly set a password for the bulk upload but there's the option to leave it blank in which case it won't set it.

File: src/test/scala/uk/gov/nationalarchives/keycloak/users/ApiLambdaSpec.scala [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+    val event = new APIGatewayV2HTTPEvent()
+    val requestContext = new RequestContext()
+    val http = new Http()
+    http.setMethod("POST")
+    requestContext.setHttp(http)
+    event.setRequestContext(requestContext)
+    event.setBody(body)
+    event
+  }
+  "handleRequest" should "return the user ID from Keycloak" in {
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/keycloak/users/ApiLambdaSpec.scala:54]: Yeah that makes sense, I've changed it.

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+      userRepresentation.setUsername(userCredentials.userName)
+      userRepresentation.setFirstName(userCredentials.firstName)
+      userRepresentation.setLastName(userCredentials.lastName)
+      userRepresentation.setEmail(userCredentials.email)
+      userRepresentation.setEnabled(true)
+      userRepresentation.setCredentials(creds)
+
+      val bodyUserGroups = userCredentials.body.map(value => s"/transferring_body_user/$value").toList
+      val userTypeGroups = userCredentials.userType.map(value => s"/user_type/${value}_user").toList
```
[the remark this answers is not in the document]
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala:60]: They're not required. If they're missing then the list of groups is empty and the Keycloak API is happy with that.

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala (hunk 1 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/keycloak/users/KeycloakUsers.scala:27]: It is quite unlikely but I've set it in application.conf with an optional environment variable and a default of tdr. That way if we do want to upload to the master realm we can add the environment variable.

File: src/main/scala/uk/gov/nationalarchives/keycloak/users/LambdaRunner.scala (hunk 2 above)
Comment (MancunianSam) [reply src/main/scala/uk/gov/nationalarchives/keycloak/users/LambdaRunner.scala:18]: This is just code so people can run it outside of the Lambda for debugging, it's not called by the lambda for anything. It's similar to this one https://github.com/nationalarchives/tdr-notifications/blob/master/src/main/scala/uk/gov/nationalarchives/notifications/LambdaRunner.scala 
The bucket should be in the sandbox though so we're not messing with integration. I've changed the bucket.

**Review by TomJKing — APPROVED** [review]

File: src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+    val requests: List[TestUserRequest] = lambdaSpecUtils.userCreateCalls.map(r => lambdaSpecUtils.testRequest(r.getRequest.getBodyAsString))
+    for {
+      x <- 1 until 4
+    } yield {
+      val userOpt = requests.find(_.email == s"test$x@test.com")
+      userOpt.isDefined should equal(true)
+      val user = userOpt.get
+      user.firstName should equal(s"First$x Name$x")
+      user.lastName should equal(s"Last$x Name$x")
+      user.lastName should equal(s"Last$x Name$x")
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala:164]: Repeated check

File: src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+    new CSVLambda().handleRequest(s3Event, null)
+
+    val requests: List[TestUserRequest] = lambdaSpecUtils.userCreateCalls.map(r => lambdaSpecUtils.testRequest(r.getRequest.getBodyAsString))
+    for {
+      x <- 1 until 4
+    } yield {
+      val userOpt = requests.find(_.email == s"test$x@test.com")
+      userOpt.isDefined should equal(true)
+      val user = userOpt.get
+      user.firstName should equal(s"First$x Name$x")
```
Comment (TomJKing) [inline src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala:162]: Should check the username is the user's email address

**Review by TomJKing — COMMENTED** [review]
Looks good. Just couple of minor comments.

File: src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala (hunk 12 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala:162]: Good point, I've added in a check for it.

File: src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala (hunk 11 above)
Comment (MancunianSam) [reply src/test/scala/uk/gov/nationalarchives/keycloak/users/CSVLambdaSpec.scala:164]: I've taken the duplicate one out.

**Review by TomJKing — APPROVED** [review]
Looks good.

**Pull request merged**